pkg/ipcache/kvstore.go

```diff
@@ -60,7 +121,43 @@ func upsertToKVStore(ipKey string, ipIDPair identity.IPIdentityPair) error {
 		logfields.Modification: Upsert,
 	}).Debug("upserting IP->ID mapping to kvstore")
 
-	return kvstore.Update(ipKey, marshaledIPIDPair, true)
+	r.Lock()
+	defer r.Unlock()
+	refcnt, ok := r.keys[ipKey]
+	if ok {
+		refcnt++
+	} else {
+		refcnt = 1
+	}
+	err = r.store.upsert(ipKey, marshaledIPIDPair, true)
+	if err == nil {
+		r.keys[ipKey] = refcnt
+	}
+	return err
+}
+
+// release removes a reference to the specified key. If the number of
+// references reaches 0, the key is removed from the underlying kvstore.
+func (r *kvReferenceCounter) release(key string) (err error) {
+	r.Lock()
+	defer r.Unlock()
+
+	refcnt, ok := r.keys[key]
+	if ok {
+		refcnt--
+	} else {
+		refcnt = 0
```

Comment:
That's not OK. You'll try to release they key from the store in this case, and I'm not sure whether that will be ok.
Instead, I think you should panic in this case, since this is a bug.

Reply:
I believe this will be okay, it will be similar to the behaviour found in #4570 - the key won't be found, the delete will fail then we'll pass the error back up to the caller and it will be logged.

Panic sounds drastic considering the consequences. I'm open to adding this to master, but it would make me nervous about whether we want to backport this PR any time soon unless we give it some good soak time.

Also, this will be dismissed by addressing your feedback immediately above.

Reply:
I'm absolutely against panic for anything that can reoccur. It' impossible to work around and leads to a total outage of the entire cluster if it happens based on kvstore state.

Panic is only acceptable if we can't continue and are potentially corrupting data.

Panic is not OK to force developers and users to pay attention. We should find a way to monitor errors and resolve them.